onnxruntime/core/providers/cuda/math/topk.cc

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "topk.h"
+#include "topk_impl.h"
+// #include "core/providers/cuda/cuda_allocator.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+ONNX_OPERATOR_KERNEL_EX(
+  TopK,
+  kOnnxDomain,
+  11,
+  kCudaExecutionProvider,
+  KernelDefBuilder().TypeConstraint("T", DataTypeImpl::AllFixedSizeTensorTypes()),
+  TopK);
+
+TopK::TopK(const OpKernelInfo& info) : CudaKernel(info) {
+  info.GetAttrOrDefault<int64_t>("axis",    &axis_,   -1);
+  info.GetAttrOrDefault<int64_t>("largest", &largest_, 1);
+  info.GetAttrOrDefault<int64_t>("sorted",  &sorted_,  1);
+}
+
+#define ISTYPE(T)   tensor_X->DataType() == DataTypeImpl::GetType<T>()
+#define TOPKIMPL(T) TopKImpl<T> (tensor_X->Data<T>(),                               \
+                                 static_cast<T*>(tensor_V->MutableDataRaw()),       \
+                                 static_cast<int64_t*>(tensor_I->MutableDataRaw()), \
+                                 elem_nums_cuda.GpuPtr(),                           \
+                                 elem_nums.size(),                                  \
+                                 axis, K, largest_, sorted_, N, dimension)
+
+Status TopK::ComputeInternal(OpKernelContext* ctx) const {
+  auto tensor_X = ctx->Input<Tensor>(0);
+  ORT_ENFORCE(nullptr != tensor_X);
+  auto rank = static_cast<int64_t>(tensor_X->Shape().NumDimensions());
+  auto axis = axis_ < 0 ? rank + axis_ : axis_;
+  ORT_ENFORCE(axis > -1 && axis < rank);
+
+  auto tensor_K = ctx->Input<Tensor>(1);
+  ORT_ENFORCE(nullptr != tensor_K);
+  int64_t K = 0;
+  cudaMemcpyAsync(&K, tensor_K->Data<int64_t>(), sizeof(int64_t), cudaMemcpyDeviceToHost);
```

Comment:
You can call .OutputMemoryType<OrtMemTypeCPUOutput>(1) in kernel register to force tensor_K as a cpu output.

Reply:
Line removed.